src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,903 @@
+/**
+ * @file feature_extraction_impl.hpp
+ * @author Nilay Jain
+ *
+ * Implementation of feature extraction methods.
+ */
+#ifndef MLPACK_METHODS_EDGE_BOXES_STRUCTURED_TREE_IMPL_HPP
+#define MLPACK_METHODS_EDGE_BOXES_STRUCTURED_TREE_IMPL_HPP
+
+
+#include "feature_extraction.hpp"
+#include <map>
+
+namespace mlpack {
+namespace structured_tree {
+
+template<typename MatType, typename CubeType>
+StructuredForests<MatType, CubeType>::
+StructuredForests(const std::map<std::string, int>& inMap)
+{
+  this->options = inMap;
+}
+
+template<typename MatType, typename CubeType>
+MatType StructuredForests<MatType, CubeType>::
+LoadData(MatType& images, MatType& boundaries, MatType& segmentations)
+{
+  int num_images = this->options["num_images"];
+  int row_size = this->options["row_size"];
+  int col_size = this->options["col_size"];
+  MatType input_data(num_images * row_size * 5, col_size);
+  // we store the input data as follows: 
+  // images (3), boundaries (1), segmentations (1).
+  int loop_iter = num_images * 5;
+  size_t row_idx = 0;
+  int col_i = 0, col_s = 0, col_b = 0;
+  for(size_t i = 0; i < loop_iter; ++i)
+  {
+    if (i % 5 == 4)
+    {
+      input_data.submat(row_idx, 0, row_idx + row_size - 1,\
+        col_size - 1) = MatType(segmentations.colptr(col_s),\
+                                  col_size, row_size).t();
+      ++col_s;
+    }
+    else if (i % 5 == 3)
+    {
+      input_data.submat(row_idx, 0, row_idx + row_size - 1,\
+        col_size - 1) = MatType(boundaries.colptr(col_b),\
+                                  col_size, row_size).t();
+      ++col_b;
+    }
+    else
+    {
+      input_data.submat(row_idx, 0, row_idx + row_size - 1,\
+        col_size - 1) = MatType(images.colptr(col_i),
+                                  col_size, row_size).t();
+      ++col_i;  
+    }
+    row_idx += row_size;
+  }
+  return input_data;
+} 
+
+template<typename MatType, typename CubeType>
+arma::vec StructuredForests<MatType, CubeType>::
+GetFeatureDimension()
+{
+  /*
+  shrink: amount to shrink channels
+  p_size: size of image patches
+  n_cell: number of self similarity cells
+  n_orient: number of orientations per gradient scale
+  */
+  arma::vec P(2);
+  int shrink, p_size, n_cell;
+  shrink = this->options["shrink"];
+  p_size = this->options["p_size"];
+  n_cell = this->options["n_cell"];
+
+  /*
+  n_color_ch: number of color channels
+  n_grad_ch: number of gradient channels
+  n_ch: total number of channels
+  */
+  int n_color_ch, n_grad_ch, n_ch;
+  if (this->options["rgbd"] == 0)
+    n_color_ch = 3;
+  else
+    n_color_ch = 4;
+
+  n_grad_ch = 2 * (1 + this->options["n_orient"]);
+
+  n_ch = n_color_ch + n_grad_ch;
+  P[0] = pow((p_size / shrink) , 2) * n_ch;
+  P[1] = pow(n_cell , 2) * (pow (n_cell, 2) - 1) / 2 * n_ch;
+  return P;
+}
+
+template<typename MatType, typename CubeType>
+arma::vec StructuredForests<MatType, CubeType>::
```

Comment:
Since this function do not alter any data member, call non-const member function, I think it is a good candidate to declare as const member function.